src/hotspot/share/classfile/systemDictionary.cpp

```diff
@@ -1615,26 +1611,32 @@ InstanceKlass* SystemDictionary::find_or_define_instance_class(Symbol* class_nam
 // is held, to ensure that the compiler is not using the class hierarchy, and that deoptimization will kick in
 // before a new class is used.
 
-void SystemDictionary::add_to_hierarchy(InstanceKlass* k) {
-  assert(k != nullptr, "just checking");
-  if (Universe::is_fully_initialized()) {
-    assert_locked_or_safepoint(Compile_lock);
-  }
+void SystemDictionary::add_to_hierarchy(InstanceKlass* k, JavaThread* current) {
```

Comment:
Since you're changing add_to_hierarchy so much, I think you might as well move it to InstanceKlass also.

These changes look good. Removing the Compile_lock around dictionary update can be done later (at at lower priority).

Reply:
Hi Coleen, so I looked at moving it to IK.
There will be some additional changes, some method may now be private instead, etc...
You think this should be done here?

Reply:
ok, maybe it's ok where it is and we can move it with a future RFE.

Reply:
great